### DISCUSSION TITLE
Some tests Gamestream vs Sunshine, Windows vs Linux

### DISCUSSION THREAD

**Discussion by KuleRucket** [LizardByte/community#110] [Sunshine]
I just ran some tests with different hosts, all streaming at my usual 1440p/120 fps. I don't have a high speed camera so haven't taken measurements. To me the differences are barely perceivable.

Test 1 - Gamestream/Windows
https://youtu.be/c1ybNl7XmDg

Test 2 - Sunshine/Windows
https://youtu.be/6xKzH5vgpoA

Test 3 - Sunshine/Linux (NvFBC + nvenc p3)
https://youtu.be/Xa1A5T_Pn_I

**Comment by marmite22** [comment]
How do you display those stats?

**Reply by KuleRucket** [reply]
ctrl-alt-shift-s

**Comment by paju1986** [comment]
can you include a stopwatch on screen so we can see really what is the latency in ms?
Like i did on here:
https://www.reddit.com/r/MoonlightStreaming/comments/zowf2w/sunshine_moonlight_latency_test/

**Comment by KuleRucket** [comment]
I retested with a similar timer and the results are quite interesting. This time both Windows and Linux sunshine are running the nightly version from yesterday.

[Gamestream to Moonlight](https://www.youtube.com/watch?v=A40Zz9J7bg0): ~44-47ms
[Sunshine (Windows) to Moonlight](https://www.youtube.com/watch?v=CiACfMWd7vQ): ~46-50ms
[Sunshine (Linux) to Moonlight](https://www.youtube.com/watch?v=1pC8EhAKtTs): 40ms

EDIT and for the hell of it:
[Parsec](https://www.youtube.com/watch?v=37Nqe7m8B50): ~45ms
I'm quite surprised by Parsec as it always felt much worse in games.

**Reply by KuleRucket** [reply]
I'm pretty sure that most of the delay is in the latency of the screen on the laptop. It's just an office Notebook that was the easiest thing to move near my host PC.

**Reply by ReenigneArcher** [reply]
Parsec is 720p though and a lower bitrate, I'd image?

**Reply by KuleRucket** [reply]
Parsec was also 1440p/120 with a 50M max birate. Although it was only using about 13Mbit so could have been self throttling. Parsec timestamp varied a lot more

**Comment by paju1986** [comment]
You got the exact same results as me on Linux to Linux

**Comment by DHNishi** [comment]
Just to toss my data points in, I'm observing substantially worse performance on Windows with a 13700K+4090 build, streaming at 1080p 120 FPS.

[GFE](https://imgur.com/a/L0MkpU4) 1-3 frames behind @ 1080p (8ms to 25ms)

vs

[Sunshine](https://imgur.com/a/lG0HSAh) 19-21 frames behind @ 1080p (158ms to 175ms)

Despite the fact that the BlurBusters test says "60Hz", it appears to be updating at 120Hz so this is like 167ms for Sunshine (Windows) to Moonlight. Sunshine is also highly inconsistent because I can't get it to hold more than 80FPS without choking (this is true for all the AVC NVENC, HEVC NVENC, and libx264 encoders) and the frame timings are all over the place while GFE holds a stable 120.

Parsec runs with a similar latency to GFE, maybe slightly slower. I haven't measured Steam Link but it certainly feels more responsive than Sunshine.

Those tests were run with the following Sunshine settings:
Latest dev daily build
Run app as admin, not as service
Both client and server at 120hz refresh rate.

Nvenc preset: llhq (does not seem to make a big difference)
Nvenc rate control: auto
Nvenc encoder: auto 

HEVC Support off (HEVC seems to perform worse for me)
Sunshine will specify based on encoder
Force encoder: NVENC
FEC: 20 (did not seem to make a difference)

Dwmflush: enabled (did not seem to make a difference)

I have not tested streaming from this same machine as a Linux host yet

**Reply by DHNishi** [reply]
Small update: Switching from llhq to High Performance on the Nvenc preset shaved one frame off. It's still pretty bad at 18-20 frames behind but it was a small improvement.

**Reply by KuleRucket** [reply]
Something is definitely wrong there and it should be a lot better. Unfortunately it often isn't easy discovering where the issue is.

**Reply by DHNishi** [reply]
I think I have a clue as to what could be going wrong but I haven't yet figured it out.

I installed Ubuntu 22.04 and re-ran the performance tests with Sunshine. When I streamed at 1080p@60hz, I was able to get around 6 frames of delay (6x16.67ms = 100ms) and with 1080p@120hz I was able to get 9 frames of delay (9x8.33ms = 75ms). Much better than Sunshine (Windows) but still pretty bad.

I then applied the NvFBC patch to my NVIDIA driver and then my Linux performance *regressed* to my Windows performance. I was seeing 19-21 frames of delay at 120hz on the Linux side. For some reason, screencasting with X11 runs much better on my machine.

**Reply by KuleRucket** [reply]
This sort of behaviour is infuriating. I previously added some timer calculations to the code and measured nvfbc::snapshot(). A snapshot via NvFBC + copy to system memory was taking ~50 microseconds.

**Reply by DHNishi** [reply]
One more performance update: 

4090 (Windows) Sunshine => 2070 Moonlight-QT amazing (+~16ms)
4090 (Windows) Sunshine => Moonlight-Android bad (+~70ms)
4090 (Windows) Sunshine => Moonlight-iOS bad (+~70ms)
4090 (Windows) Sunshine => Moonlight-Chrome terrible (+\~ 180ms)

4090 (Linux) Sunshine => 2070 Moonlight-QT untested
4090 (Linux) Sunshine => Moonlight-Android bad (+~70ms)
4090 (Linux) Sunshine => Moonlight-iOS bad (+~70ms)
4090 (Linux) Sunshine => Moonlight-Chrome bad (NvFBC is terrible) (+\~ 100ms, \~170ms for NvFBC)

2070 (Windows) Sunshine => 4090 Moonlight-QT OK (+~33ms)
2070 (Windows) Sunshine => Moonlight-Android bad (+~70ms)
2070 (Windows) Sunshine => Moonlight-iOS bad (+~70ms)
2070 (Windows) Sunshine => Moonlight-Chrome very bad (+\~ 125ms)

4090 GFE => Moonlight-* amazing (+~16-25ms) 
4090 Parsec => Windows Parsec great (+~16-33ms)

**Reply by paju1986** [reply]
How did you perform the measurements?

**Reply by DHNishi** [reply]
I used https://www.testufo.com/flicker and placed the two displays next to each other with it full screened. I then used the "Is it Snappy?" app to get some 240hz recordings to check the difference in the number of frames displayed on the screen.

**Comment by mdhen** [comment]
I'm not comparing on my own hardware, but I don't see any linux/amd here so I thought I would chime in:

RX 6800XT (Linux) Sunshine using vaapi => WebOS moonlight-tv using NDL. Desktop to router is hard wired, then router to TV is wifi, only a few feet away.

at 4k, running 80 to 100 FPS, Network RTT is 3ms (var 3ms) - and decode time is between 0.2 and 0.5 ms

[a comment by an automated account is not shown]
